Kernel/Process.cpp

```diff
@@ -763,19 +763,35 @@ pid_t Process::sys$fork(RegisterState& regs)
 int Process::do_exec(NonnullRefPtr<FileDescription> main_program_description, Vector<String> arguments, Vector<String> environment, RefPtr<FileDescription> interpreter_description)
 {
     ASSERT(is_ring3());
+
     auto path = main_program_description->absolute_path();
     dbg() << "do_exec(" << path << ")";
-    // FIXME(Thread): Kill any threads the moment we commit to the exec().
+
     if (thread_count() != 1) {
-        dbgprintf("Gonna die because I have many threads! These are the threads:\n");
         for_each_thread([](Thread& thread) {
-            dbgprintf("Thread{%p}: TID=%d, PID=%d\n", &thread, thread.tid(), thread.pid());
+            if (&thread == Thread::current || thread.is_dead())
+                return IterationDecision::Continue;
+
+            dbgprintf("Will do an exec, killing thread with TID=%d, state=%d\n",
+                      thread.tid(), thread.state());
+
+            thread.set_should_die();
+
+            // FIXME: Shouldn't thread.set_should_die always set the state to dying?
+            // Right now it ignores threads which are in kernel and threads which are blocked
+            if (!thread.is_dead())
+                thread.set_state(Thread::State::Dying);
+
             return IterationDecision::Continue;
         });
-        ASSERT(thread_count() == 1);
-        ASSERT_NOT_REACHED();
     }
 
+    int tries = thread_count() * 4;
+    while (thread_count() > 1 && --tries)
+        Scheduler::yield();
```

Comment:
This will kill all threads in a program that calls `exec()` with bad arguments. In such a case, `exec()` should simply fail and keep the program running like before.

Also, we should do this without having to wait for threads to die. As long as we've marked the threads for death and made them non-runnable, there's no need for us to wait for them to get deallocated.

Reply:
Yeah, I spent some time to figure things out. I just removed that part and remembered why I put it there: `Queued` threads won't exit (running, blocked, etc. are fine) and they will hit an assertion in `Kernel::Thread::wake_from_queue()`.

Reply:
Figured it out (at least I think I did). I amended the previous commit, now it won't need to actually wait for threads to finish. My only question remains if the `m_joiner` should also be set to null inside the threads. Makes sense to clean it, since the process which does the exec isn't the joiner any more.

Reply:
Sure, nobody is going to join these threads on successful `exec()` so let's clear the `m_joiner`.

Reply:
Set that too. This *should* cover all the cases. In my tests (sleeping threads, running threads, queued threads), it finishes them nicely.

Reply:
Also, I investigated `exit` too, and that doesn't really work if you have queued threads, the program will exit, but it will just hang the terminal.

I pushed a working solution for both exec and exit & updated the PR.